### PR TITLE
crypto:  add sr25519 signature scheme

### PR DESCRIPTION
**Pull request by sunnya97**
[sr25519](https://github.com/w3f/schnorrkel) is the schnorrs on ristretto signature scheme that's primarily used by [Parity Substrate](https://substrate.io/).  Adding to Tendermint will be useful for cross-chain compatibility purposes like IBC and hard spoons of Substrate-based chains.  This makes use of the [go-schnorrkel](https://github.com/ChainSafe/go-schnorrkel) library written by Chainsafe.

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: crypto/sr25519/sr25519_test.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/sr25519"
+)
+
+func TestSignAndValidateEd25519(t *testing.T) {
```
Comment (marbar3778) [inline crypto/sr25519/sr25519_test.go]:
```suggestion
func TestSignAndValidateSr25519(t *testing.T) {
```

File: crypto/sr25519/sr25519.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
+package sr25519
+
+import (
+	"bytes"
+	"crypto/subtle"
+	"fmt"
+	"io"
+
+	amino "github.com/tendermint/go-amino"
+
```
Comment (tessr) [inline crypto/sr25519/sr25519.go]:
```suggestion
```

File: crypto/sr25519/sr25519.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+)
+
+//-------------------------------------
+
+var _ crypto.PrivKey = PrivKeySr25519{}
+
+const (
+	PrivKeyAminoName = "tendermint/PrivKeySr25519"
+	PubKeyAminoName  = "tendermint/PubKeySr25519"
+	// Size of an Edwards25519 signature. Namely the size of a compressed
```
Comment (tessr) [inline crypto/sr25519/sr25519.go]:
```suggestion

	// SignatureSize is the size of an Edwards25519 signature. Namely the size of a compressed
```

File: crypto/sr25519/sr25519.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// if it's derived from user input.
+func GenPrivKeyFromSecret(secret []byte) PrivKeySr25519 {
+	seed := crypto.Sha256(secret) // Not Ripemd160 because we want 32 bytes.
+	var bz [32]byte
+	copy(bz[:], seed)
+	privKey, _ := schnorrkel.NewMiniSecretKeyFromRaw(bz)
+	return privKey.ExpandEd25519().Encode()
+}
+
+//-------------------------------------
```
Comment (tessr) [inline crypto/sr25519/sr25519.go]: I know these ascii separators are common in this repo, but I don't understand why we use them. Why not break domains into separate files (in the same package)? That would be a little more editor/merge conflict friendly as well.

File: crypto/sr25519/sr25519.go (hunk 3 above)
Comment (marbar3778) [inline crypto/sr25519/sr25519.go]:
```suggestion
	// Size of an Sr25519 signature. Namely the size of a compressed
```

File: crypto/sr25519/sr25519.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+var _ crypto.PrivKey = PrivKeySr25519{}
+
+const (
+	PrivKeyAminoName = "tendermint/PrivKeySr25519"
+	PubKeyAminoName  = "tendermint/PubKeySr25519"
+	// Size of an Edwards25519 signature. Namely the size of a compressed
+	// Edwards25519 point, and a field element. Both of which are 32 bytes.
```
Comment (marbar3778) [inline crypto/sr25519/sr25519.go]:
```suggestion
	// Sr25519 point, and a field element. Both of which are 32 bytes.
```

File: crypto/sr25519/sr25519.go (hunk 4 above)
Comment (melekes) [reply crypto/sr25519/sr25519.go]: https://github.com/tendermint/tendermint/issues/3619

Comment (tessr) [reply crypto/sr25519/sr25519.go]: Great. Let's make sure this PR gets reorganized before landing it.

**Comment by sunnya97** [comment]
@tessr The primary reason I'm interested in this tbh, is because of a project I work on called [Straightedge](https://straighted.ge/).  It's a fork of a Substrate-based chain called [Edgeware](https://edgewa.re/) ([with some changes to the distribution](https://straighted.ge/faq/)).  I'm trying to transition Straightedge from Substrate chain to the Cosmos SDK, but all of the the keys from the [Edgeware lockdrop](https://medium.com/commonwealth-labs/whats-in-a-lockdrop-194218a180ca) were all sr25519 keys, and so needed to add support for the key type.

File: crypto/sr25519/privkey.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	return sigBytes[:], nil
+}
+
+// PubKey gets the corresponding public key from the private key.
+func (privKey PrivKeySr25519) PubKey() crypto.PubKey {
+
+	secretKey := &(schnorrkel.SecretKey{})
+	err := secretKey.Decode(privKey)
+	if err != nil {
+		panic("invalid private key")
```
Comment (melekes) [inline crypto/sr25519/privkey.go]:
```suggestion
		panic(fmt.Sprintf("Invalid private key: %v", err))
```

File: crypto/sr25519/privkey.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	"crypto/subtle"
+	"io"
+
+	"github.com/tendermint/tendermint/crypto"
+
+	schnorrkel "github.com/ChainSafe/go-schnorrkel"
+)
+
+// PrivKeySr25519 implements crypto.PrivKey.
+type PrivKeySr25519 [32]byte
```
Comment (melekes) [inline crypto/sr25519/privkey.go]: do you think it makes sense to create `PrivKeySr25519Size = 32` const and use it throughout the code?

File: crypto/sr25519/pubkey.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func (pubKey PubKeySr25519) String() string {
+	return fmt.Sprintf("PubKeySr25519{%X}", pubKey[:])
+}
+
+// nolint: golint
+func (pubKey PubKeySr25519) Equals(other crypto.PubKey) bool {
+	if otherEd, ok := other.(PubKeySr25519); ok {
+		return bytes.Equal(pubKey[:], otherEd[:])
```
Comment (melekes) [inline crypto/sr25519/pubkey.go]: not a constant time compare?

File: crypto/sr25519/sr25519_test.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package sr25519_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/crypto"
```
Comment (melekes) [inline crypto/sr25519/sr25519_test.go]:
```suggestion

	"github.com/tendermint/tendermint/crypto"
```

**Comment by marbar3778** [comment]
does this need to be updated as well? https://github.com/tendermint/tendermint/blob/master/types/protobuf.go

also can you update the readme in the crypto package: https://github.com/tendermint/tendermint/blob/master/crypto/README.md

also this needs updating: https://github.com/tendermint/spec/blob/master/spec/blockchain/encoding.md#public-key-cryptography

File: crypto/sr25519/pubkey.go (hunk 8 above)
Comment (sunnya97) [reply crypto/sr25519/pubkey.go]: Idk, this was the way we did it in the Ed25519 version.  https://github.com/tendermint/tendermint/blob/master/crypto/ed25519/ed25519.go#L164

**Comment by jackzampolin** [comment]
Not having an amino dependency here would be ideal.

File: crypto/sr25519/pubkey.go (hunk 8 above)
Comment (melekes) [reply crypto/sr25519/pubkey.go]: probably not needed

File: crypto/sr25519/pubkey.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	return publicKey.Verify(signature, signingContext)
+}
+
+func (pubKey PubKeySr25519) String() string {
+	return fmt.Sprintf("PubKeySr25519{%X}", pubKey[:])
+}
+
+// nolint: golint
```
Comment (melekes) [inline crypto/sr25519/pubkey.go]: please remove

**Comment by marbar3778** [comment]
just remembered that it would be nice to have a changelog entry, mainly for the people that keep up with changelog to see what is new.

**Comment by tessr** [comment]
@sunnya97 thank you for the context! appreciated.

**Comment by marbar3778** [comment]
don't merge please, this is blocked on https://github.com/ChainSafe/go-schnorrkel/pull/5

**Comment by sunnya97** [comment]
https://github.com/ChainSafe/go-schnorrkel/pull/5 merged and Changelog updated 👍

**Comment by melekes** [comment]
At some point we will need to find a way for applications to inject their own keys, which does not require adding them to Tendermint Core.

**Comment by marbar3778** [comment]
> At some point we will need to find a way for applications to inject their own keys, which does not require adding them to Tendermint Core.

The first step in doing this would be to get rid of init functions in the `codec.go` files

**Comment by melekes** [comment]
> The first step in doing this would be to get rid of init functions in the codec.go files, correct?

yes

**Pull request merged**